cachito/web/models.py

```diff
@@ -186,6 +187,32 @@ def to_purl(self):
             # see https://github.com/package-url/purl-spec/issues/63 for further reference
             purl_name = self.name.replace("/", "%2F")
             return f"pkg:golang/{purl_name}@{self.version}"
+
+        elif self.type == "npm":
+            purl_name = self.name.replace("@", "%40")
+            match = re.match(r"(?P<protocol>[^:]+):(?://)?(?P<suffix>.+)", self.version)
+            if not match:
+                return f"pkg:npm/{purl_name}@{self.version}"
+            protocol = match.group("protocol")
+            suffix = match.group("suffix")
+            if protocol in ("github", "gitlab", "bitbucket"):
+                # github:namespace/name#ref or gitlab:ns1/ns2/name#ref
+                match_forge = re.match(
+                    r"(?P<namespace>.+)/(?P<name>[^#/]+)#(?P<version>.+)$", suffix
+                )
+                if not match_forge:
+                    raise ValueError(f"Could not convert version {self.version} to purl")
+                forge = match_forge.groupdict()
+                return f"pkg:{protocol}/{forge['namespace']}/{forge['name']}@{forge['version']}"
+            elif protocol in ("git", "git+http", "git+https", "git+ssh"):
+                qualifier = urllib.parse.quote(self.version, safe="")
```

Comment:
Interestingly, the purl-spec repo does not "quote" this:
https://github.com/package-url/purl-spec

Reply:
It does say that for qualifiers, "A value must be a percent-encoded string"

Reply:
It's interesting that the examples don't show that. It makes sense to me though.